mountpoint-s3/src/fs.rs

```diff
@@ -76,13 +77,14 @@ impl<Client: ObjectClient> UploadState<Client> {
             Err(e) => {
                 error!("write failed: {e}");
                 match e {
-                    UploadWriteError::PutRequestFailed(_) => {
+                    UploadWriteError::PutRequestFailed(_) | UploadWriteError::ObjectTooBig { .. } => {
```

Comment:
Arguably, we could treat `ObjectTooBig` like `OutOfOrderWrite`: do not set the upload to `Failed` and allow for retries with a size that fits and/or subsequent fsync/release.
Is this a conscious choice not to do that?

I may also not fully understand the issue with kernel retries: wouldn't they also fail just with the size check in `UploadRequest::write`?

Reply:
But I do like recording the error in `Failed` in general!

Reply:
It was a conscious choice because I'm worried about uploading truncated objects. On reflection I think we should do the same thing for out-of-order writes -- we'd rather fail loudly in case applications don't actually check the results of `write`. Let's do that in a separate change.

> I may also not fully understand the issue with kernel retries: wouldn't they also fail just with the size check in UploadRequest::write?

No, because `UploadRequest::write` checks whether the upload has failed before calling `upload.write`, which is where the size check is.

Reply:
> No, because `UploadRequest::write` checks whether the upload has failed before calling `upload.write`, which is where the size check is.

Sure, but it wouldn't if we treated it like `OutOfOrderWrite`. Anyway, happy to change the behavior of the latter.